qutebrowser/misc/sql.py

```diff
@@ -0,0 +1,240 @@
+        self._name = name
+
+        constraints = constraints or {}
+        column_defs = ['{} {}'.format(field, constraints.get(field, ''))
+                       for field in fields]
+        q = Query("CREATE TABLE IF NOT EXISTS {} ({})"
+                  .format(name, ','.join(column_defs)))
+
+        q.run()
+
+    def create_index(self, name, field):
+        """Create an index over this table.
+
+        Args:
+            name: Name of the index, should be unique.
+            field: Name of the field to index.
+        """
+        q = Query("CREATE INDEX IF NOT EXISTS {} ON {} ({})"
+                  .format(name, self._name, field))
+        q.run()
+
+    def __iter__(self):
+        """Iterate rows in the table."""
+        q = Query("SELECT * FROM {}".format(self._name))
+        q.run()
+        return iter(q)
+
+    def contains_query(self, field):
+        """Return a prepared query that checks for the existence of an item.
+
+        Args:
+            field: Field to match.
+        """
+        return Query("SELECT EXISTS(SELECT * FROM {} WHERE {} = :val)"
+                     .format(self._name, field))
+
+    def __len__(self):
+        """Return the count of rows in the table."""
+        q = Query("SELECT count(*) FROM {}".format(self._name))
+        q.run()
+        return q.value()
+
+    def delete(self, value, field):
+        """Remove all rows for which `field` equals `value`.
+
+        Args:
+            value: Key value to delete.
+            field: Field to use as the key.
+
+        Return:
+            The number of rows deleted.
+        """
+        q = Query("DELETE FROM {} where {} = :val".format(self._name, field))
+        q.run(val=value)
+        if not q.numRowsAffected():
+            raise KeyError('No row with {} = "{}"'.format(field, value))
+        self.changed.emit()
+
+    def _insert_query(self, values, replace):
+        params = ','.join(':{}'.format(key) for key in values)
+        verb = "REPLACE" if replace else "INSERT"
+        return Query("{} INTO {} values({})".format(verb, self._name, params))
+
+    def insert(self, values, replace=False):
+        """Append a row to the table.
+
+        Args:
+            values: A dict with a value to insert for each field name.
+            replace: If set, replace existing values.
+        """
+        q = self._insert_query(values, replace)
+        q.run(**values)
+        self.changed.emit()
+
+    def insert_batch(self, values, replace=False):
+        """Performantly append multiple rows to the table.
+
+        Args:
+            rows: A list of lists, where each sub-list is a row.
+            values: A dict with a list of values to insert for each field name.
+        """
+        q = self._insert_query(values, replace)
+        for key, val in values.items():
+            q.bindValue(':{}'.format(key), val)
+
+        db = QSqlDatabase.database()
+        db.transaction()
+        if not q.execBatch():
```

Comment:
Hmm, you're right - now that I look at it again, I think it *should* actually work with named parameters... Still, this is the only thing I can imagine breaking stuff (changing `execBatch` to named parameters in daefa75), because this definitely worked fine recently.